lib/phoenix_live_view/diff.ex

```diff
@@ -567,15 +628,166 @@ defmodule Phoenix.LiveView.Diff do
     end)
   end
 
-  defp traverse_comprehension(dynamics, pending, components, template) do
-    Enum.map_reduce(dynamics, {pending, components, template}, fn rendereds, acc ->
-      Enum.map_reduce(rendereds, acc, fn rendered, {pending, components, template} ->
-        {diff, _, pending, components, template} =
-          traverse(rendered, {nil, %{}}, pending, components, template, false)
+  defp traverse_keyed(
+         entries,
+         previous_prints,
+         pending,
+         components,
+         template,
+         changed?,
+         stream?,
+         has_key?
+       ) do
+    diff = %{}
+    new_prints = %{}
+
+    {{diff, count, new_prints, pending, components, template, _canonical_print}, _seen_keys} =
+      Enum.reduce(
+        entries,
+        {{diff, 0, new_prints, pending, components, template, nil}, MapSet.new()},
+        fn
+          {key, vars, render},
+          {{_diff, index, _new_prints, _pending, _components, _template, _canonical_print} = acc,
+           seen_keys} ->
+            {key, seen_keys} =
+              if has_key? do
+                if MapSet.member?(seen_keys, key) do
+                  raise "found duplicate key #{inspect(key)} in comprehension"
+                else
+                  {key, MapSet.put(seen_keys, key)}
+                end
+              else
+                # no need to check for duplicates if we use the index
+                {index, seen_keys}
+              end
```

Comment:
If you prefer less nesting.

```suggestion
              cond do
                not has_key? ->
                  {index, seen_keys}

                MapSet.member?(seen_keys, key) ->
                  raise "found duplicate key #{inspect(key)} in comprehension"

                true ->
                  {key, MapSet.put(seen_keys, key)}
              end
```